### ISSUE TITLE
Inference custom model

### ISSUE THREAD

**Issue by SantaDiver** [Tencent/TurboTransformers#63]
Hello! I have Pytorch model that consists of Hugging Face BERT model followed by several PyTorch nn layers (such as Feed Forward and LSTM). Can I accelerate and run it (or at least BERT part of it) using TurboTransformers?

And how can I take advantage of variable length support? Should I just pad my tensor up to the length of the longest element?

**Comment by SantaDiver** [comment]
Is this a proper way?
```class BertWithLstm(nn.Module):
    def __init__(self, tt_bert, lstm):
        super().__init__()
        self.bert = tt_bert
        self.lstm = lstm
        self.switch = switch
        
    def forward(self, x):
        hidden_state, pooling = self.bert(x)
        bilstm_out, hidden_state = self.lstm(hidden_state)
        
        return bilstm_out

bert_tt = turbo_transformers.BertModel.from_torch(model_bert, test_device)
BertWithLstm(bert_tt, lstm)```

**Comment by feifeibear** [comment]
Definitely you can use turbo to make your PyTorch code fast!
Turbo is able to accelerate the BERT part, and you can use PyTorch to run LSTM as a postprocessing operation of BERT.
Please see the instructions for adding a postprocessing operation, a linear layer in that case.

Turbo is used the same as Pytorch. You do not need to pad and mask inputs.

**Comment by ShenYounger** [comment]
> Definitely you can use turbo to make your PyTorch code fast!
> Turbo is able to accelerate the BERT part, and you can use PyTorch to run LSTM as a postprocessing operation of BERT.
> Please see the instructions for adding a postprocessing operation, a linear layer in that case.
> 
> Turbo is used the same as Pytorch. You do not need to pad and mask inputs.

i have a question. i want to know whyTurbo doesn't need padding? what's the mechanism？

**Comment by feifeibear** [comment]
Pytorch does not pad. So do turbo. : )
A mechanism for not padding? You have to know why there exists a mechanism for padding.
Because tensorflow 1.0 is a static graph DL framework. It have to know shape of tensors in the graph, optimizes the graph before running. No such mechanism in turbo.

**Comment by SantaDiver** [comment]
@feifeibear Thank you for your answers!
About variable length. Should it be **pack_padded_sequence** from PyTorch? As far as I understand nor Hugging face nor Turbo accepts its output. Should I do it different way?

Another question is: is fp16 supported now? README points that it is matter of near future but I saw some fp16 files in sources.
And also how can I participate and contribute in project?

**Comment by feifeibear** [comment]
About variable length you can refer to this issue #27. If you are considering about a batch of variable-length sequences, you should design a batch scheduler by your self.

We are open to new contributions. We need low-precision supports and more models, or some other features you need. We have a few documents about how to add new layers. And we will add more detailed instructions latter. https://github.com/Tencent/TurboTransformers/tree/feature/albert/turbo_transformers/layers. 
Don't hesitate to contact me about your idea.

**Comment by SantaDiver** [comment]
Now everything is clear to me! Thank you very much! Closing this issue.
